File: src/main/java/tvl/parser/Lexer.java
package tvl.parser;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import tvl.core.CompileException;
import tvl.core.Pos;

/**
 * 词法分析器：把表达式源码切分为 Token 流。
 *
 * 支持：
 *  - 空白与 SQL 风格注释（-- 行注释、&#47;* *&#47; 块注释）
 *  - 整数字面量（仅十进制）、单引号字符串（'' 转义单引号）
 *  - 标识符（字母/下划线开头），关键字大小写不敏感
 *  - 算术与比较操作符、括号
 */
public final class Lexer {

    private static final Map<String, TokenType> KEYWORDS = new HashMap<>();
    static {
        KEYWORDS.put("AND", TokenType.AND);
        KEYWORDS.put("OR", TokenType.OR);
        KEYWORDS.put("NOT", TokenType.NOT);
        KEYWORDS.put("IS", TokenType.IS);
        KEYWORDS.put("TRUE", TokenType.TRUE);
        KEYWORDS.put("FALSE", TokenType.FALSE);
        KEYWORDS.put("NULL", TokenType.NULL_KW);
    }

    private final String src;
    private int pos = 0;
    private int line = 1;
    private int col = 1;
    private final List<Token> tokens = new ArrayList<>();

    public Lexer(String src) {
        this.src = src == null ? "" : src;
    }

    public List<Token> tokenize() {
        while (pos < src.length()) {
            char c = src.charAt(pos);
            if (isWhitespace(c)) {
                advance();
            } else if (c == '-' && peek(1) == '-') {
                skipLineComment();
            } else if (c == '/' && peek(1) == '*') {
                skipBlockComment();
            } else if (Character.isDigit(c)) {
                readNumber();
            } else if (c == '\'') {
                readString();
            } else if (isIdentStart(c)) {
                readIdentOrKeyword();
            } else {
                readOperator();
            }
        }
        tokens.add(new Token(TokenType.EOF, "", new Pos(pos, line, col), pos));
        return tokens;
    }

    private void skipLineComment() {
        advance(); // -
        advance(); // -
        while (pos < src.length() && src.charAt(pos) != '\n') {
            advance();
        }
    }

    private void skipBlockComment() {
        Pos start = here();
        advance(); // /
        advance(); // *
        while (pos < src.length()) {
            if (src.charAt(pos) == '*' && peek(1) == '/') {
                advance();
                advance();
                return;
            }
            advance();
        }
        throw new CompileException("UNTERMINATED_COMMENT",
                "块注释缺少结束标记 */", start, 2);
    }

    private void readNumber() {
        Pos start = here();
        int begin = pos;
        while (pos < src.length() && Character.isDigit(src.charAt(pos))) {
            advance();
        }
        String text = src.substring(begin, pos);
        tokens.add(new Token(TokenType.INTEGER, text, start, pos));
    }

    private void readString() {
        Pos start = here();
        advance(); // 开引号
        StringBuilder sb = new StringBuilder();
        while (true) {
            if (pos >= src.length()) {
                throw new CompileException("UNTERMINATED_STRING",
                        "字符串字面量缺少结束单引号 '", start, 1);
            }
            char c = src.charAt(pos);
            if (c == '\'') {
                if (peek(1) == '\'') {
                    sb.append('\'');
                    advance();
                    advance();
                } else {
                    advance(); // 闭引号
                    break;
                }
            } else {
                sb.append(c);
                advance();
            }
        }
        tokens.add(new Token(TokenType.STRING, sb.toString(), start, pos));
    }

    private void readIdentOrKeyword() {
        Pos start = here();
        int begin = pos;
        while (pos < src.length() && isIdentPart(src.charAt(pos))) {
            advance();
        }
        String text = src.substring(begin, pos);
        TokenType kw = KEYWORDS.get(text.toUpperCase());
        if (kw != null) {
            tokens.add(new Token(kw, text, start, pos));
        } else {
            tokens.add(new Token(TokenType.IDENT, text, start, pos));
        }
    }

    private void readOperator() {
        Pos start = here();
        char c = src.charAt(pos);
        char n = peek(1);
        switch (c) {
            case '+': single(TokenType.PLUS, "+"); return;
            case '-': single(TokenType.MINUS, "-"); return;
            case '*': single(TokenType.STAR, "*"); return;
            case '/': single(TokenType.SLASH, "/"); return;
            case '(': single(TokenType.LPAREN, "("); return;
            case ')': single(TokenType.RPAREN, ")"); return;
            case '=': single(TokenType.EQ, "="); return;
            case '<':
                if (n == '=') { twoChar(TokenType.LE, "<="); }
                else if (n == '>') { twoChar(TokenType.NE, "<>"); }
                else { single(TokenType.LT, "<"); }
                return;
            case '>':
                if (n == '=') { twoChar(TokenType.GE, ">="); }
                else { single(TokenType.GT, ">"); }
                return;
            case '!':
                if (n == '=') {
                    twoChar(TokenType.NE, "!=");
                    return;
                }
                break;
            default:
                break;
        }
        throw new CompileException("UNEXPECTED_CHAR",
                "无法识别的字符 '" + c + "'（U+" + String.format("%04X", (int) c) + "）",
                start, 1);
    }

    private void single(TokenType type, String text) {
        Pos start = here();
        advance();
        tokens.add(new Token(type, text, start, pos));
    }

    private void twoChar(TokenType type, String text) {
        Pos start = here();
        advance();
        advance();
        tokens.add(new Token(type, text, start, pos));
    }

    private boolean isWhitespace(char c) {
        return c == ' ' || c == '\t' || c == '\r' || c == '\n' || c == '\f';
    }

    private boolean isIdentStart(char c) {
        return Character.isLetter(c) || c == '_';
    }

    private boolean isIdentPart(char c) {
        return Character.isLetterOrDigit(c) || c == '_';
    }

    private Pos here() {
        return new Pos(pos, line, col);
    }

    private char peek(int ahead) {
        int p = pos + ahead;
        return p < src.length() ? src.charAt(p) : '\0';
    }

    private void advance() {
        if (pos < src.length()) {
            if (src.charAt(pos) == '\n') {
                line++;
                col = 1;
            } else {
                col++;
            }
            pos++;
        }
    }
}

File: app/language/lexer.py
"""Lexer for the small scripting language (see README.md for the grammar).

The lexer is deliberately hand-written (no third-party parsing dependency) so
that the whole analysis pipeline stays auditable.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto


class TokenType(Enum):
    # literals / names
    INTEGER = auto()
    STRING = auto()
    TRUE = auto()
    FALSE = auto()
    NAME = auto()
    # keywords
    FUNC = auto()
    RETURN = auto()
    IF = auto()
    ELSE = auto()
    WHILE = auto()
    VAR = auto()
    NIL = auto()
    AND = auto()
    OR = auto()
    NOT = auto()
    # punctuation
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    COMMA = auto()
    SEMI = auto()
    ASSIGN = auto()
    # operators
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    EQ = auto()
    NE = auto()
    LT = auto()
    LE = auto()
    GT = auto()
    GE = auto()
    BANG = auto()
    EOF = auto()


KEYWORDS = {
    "func": TokenType.FUNC,
    "return": TokenType.RETURN,
    "if": TokenType.IF,
    "else": TokenType.ELSE,
    "while": TokenType.WHILE,
    "var": TokenType.VAR,
    "nil": TokenType.NIL,
    "true": TokenType.TRUE,
    "false": TokenType.FALSE,
    "and": TokenType.AND,
    "or": TokenType.OR,
    "not": TokenType.NOT,
}


@dataclass(frozen=True)
class Token:
    type: TokenType
    value: str
    line: int
    col: int


class LexError(Exception):
    def __init__(self, message: str, line: int, col: int):
        super().__init__(f"{message} at line {line}, column {col}")
        self.line = line
        self.col = col


_SIMPLE = {
    "(": TokenType.LPAREN,
    ")": TokenType.RPAREN,
    "{": TokenType.LBRACE,
    "}": TokenType.RBRACE,
    ",": TokenType.COMMA,
    ";": TokenType.SEMI,
    "+": TokenType.PLUS,
    "-": TokenType.MINUS,
    "*": TokenType.STAR,
    "/": TokenType.SLASH,
    "%": TokenType.PERCENT,
}


class Lexer:
    def __init__(self, source: str):
        self.src = source
        self.n = len(source)
        self.pos = 0
        self.line = 1
        self.col = 1

    # -- low-level helpers -------------------------------------------------
    def _peek(self, offset: int = 0) -> str:
        p = self.pos + offset
        return self.src[p] if p < self.n else ""

    def _advance(self) -> str:
        ch = self.src[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def _skip_trivia(self) -> None:
        while self.pos < self.n:
            ch = self._peek()
            if ch in " \t\r\n":
                self._advance()
            elif ch == "/" and self._peek(1) == "/":
                while self.pos < self.n and self._peek() != "\n":
                    self._advance()
            elif ch == "/" and self._peek(1) == "*":
                start_line, start_col = self.line, self.col
                self._advance()
                self._advance()
                while self.pos < self.n and not (
                    self._peek() == "*" and self._peek(1) == "/"
                ):
                    self._advance()
                if self.pos >= self.n:
                    raise LexError("unterminated block comment", start_line, start_col)
                self._advance()
                self._advance()
            else:
                return

    def _read_string(self, quote: str, line: int, col: int) -> Token:
        self._advance()  # opening quote
        chars: list[str] = []
        while self.pos < self.n and self._peek() != quote:
            ch = self._peek()
            if ch == "\n":
                raise LexError("unterminated string", line, col)
            if ch == "\\":
                self._advance()
                esc = self._peek()
                escapes = {"n": "\n", "t": "\t", "r": "\r", "\\": "\\",
                           "'": "'", '"': '"', "0": "\0"}
                if esc not in escapes:
                    raise LexError(f"invalid escape \\{esc}", self.line, self.col)
                chars.append(escapes[esc])
                self._advance()
            else:
                chars.append(self._advance())
        if self.pos >= self.n:
            raise LexError("unterminated string", line, col)
        self._advance()  # closing quote
        return Token(TokenType.STRING, "".join(chars), line, col)

    # -- main --------------------------------------------------------------
    def tokenize(self) -> list[Token]:
        tokens: list[Token] = []
        while True:
            self._skip_trivia()
            if self.pos >= self.n:
                tokens.append(Token(TokenType.EOF, "", self.line, self.col))
                return tokens
            line, col = self.line, self.col
            ch = self._peek()

            if ch.isdigit():
                start = self.pos
                while self._peek().isdigit():
                    self._advance()
                tokens.append(Token(TokenType.INTEGER, self.src[start:self.pos], line, col))
                continue
            if ch.isalpha() or ch == "_":
                start = self.pos
                while self._peek().isalnum() or self._peek() == "_":
                    self._advance()
                word = self.src[start:self.pos]
                ttype = KEYWORDS.get(word, TokenType.NAME)
                tokens.append(Token(ttype, word, line, col))
                continue
            if ch in ("'", '"'):
                tokens.append(self._read_string(ch, line, col))
                continue
            if ch == "=":
                self._advance()
                if self._peek() == "=":
                    self._advance()
                    tokens.append(Token(TokenType.EQ, "==", line, col))
                else:
                    tokens.append(Token(TokenType.ASSIGN, "=", line, col))
                continue
            if ch == "!":
                self._advance()
                if self._peek() == "=":
                    self._advance()
                    tokens.append(Token(TokenType.NE, "!=", line, col))
                else:
                    tokens.append(Token(TokenType.BANG, "!", line, col))
                continue
            if ch == "<":
                self._advance()
                if self._peek() == "=":
                    self._advance()
                    tokens.append(Token(TokenType.LE, "<=", line, col))
                else:
                    tokens.append(Token(TokenType.LT, "<", line, col))
                continue
            if ch == ">":
                self._advance()
                if self._peek() == "=":
                    self._advance()
                    tokens.append(Token(TokenType.GE, ">=", line, col))
                else:
                    tokens.append(Token(TokenType.GT, ">", line, col))
                continue
            if ch in _SIMPLE:
                self._advance()
                tokens.append(Token(_SIMPLE[ch], ch, line, col))
                continue

            raise LexError(f"unexpected character {ch!r}", line, col)
